orchestrator-salt/src/main/resources/salt/salt/post-recipes/scripts/createuserhome.sh

```diff
@@ -71,25 +71,47 @@ PATH_PREFIX="/user"
 
 export KRB5CCNAME=/tmp/krb5cc_cloudbreak_$EUID
 
+
+klist -s
+if [[ $? -ne 0 ]]; then
+  kinit_as_hdfs
+fi
+
 WEBHDFS_HTTP_POLICY=$(hdfs getconf -confkey dfs.http.policy)
 
-if [ "$WEBHDFS_HTTP_POLICY" == "HTTP_ONLY" ]; then
-  NAMENODE_HTTP_ADDRESS=$(hdfs getconf -confkey dfs.namenode.http-address)
-  WEBHDFS_URL=http://$NAMENODE_HTTP_ADDRESS/webhdfs/v1
+if ! hdfs haadmin -getAllServiceState
+then
+  echo "Single node configuration"
+  if [ "$WEBHDFS_HTTP_POLICY" == "HTTP_ONLY" ]; then
+    NAMENODE_HTTP_ADDRESS=$(hdfs getconf -confkey dfs.namenode.http-address)
+    WEBHDFS_URL=http://$NAMENODE_HTTP_ADDRESS/webhdfs/v1
+  else
+    NAMENODE_HTTP_ADDRESS=$(hdfs getconf -confkey dfs.namenode.https-address)
+    WEBHDFS_URL=https://$NAMENODE_HTTP_ADDRESS/webhdfs/v1
+  fi
 else
-  NAMENODE_HTTP_ADDRESS=$(hdfs getconf -confkey dfs.namenode.https-address)
-  WEBHDFS_URL=https://$NAMENODE_HTTP_ADDRESS/webhdfs/v1
+  echo "HA configuration"
+  NN_HOST=$(hdfs haadmin -getAllServiceState | grep active | cut -d':' -f 1)
+  echo "Active NAMENODE host: $NN_HOST"
+  if [ "$NN_HOST" != "$(hostname -f)" ]; then
+    echo "This is not the active host. Script will run on the active one"
```

Comment:
this doesn't seem like a robust approach. I have two concerns:
1) is it possible that this runs concurrently with the namenodes starting, in which case neither NN is active yet?
2) it's definitely possible that one of the NNs fails around the same time these two scripts are running. In that case either neither or both of the copies of this script will run.

it seems this script should be idempotent. How often does it get re-run? eg if for some reason we hit one of the above issues, would it run again a minute later and just be a transient problem, or would we be left in a persistent unsynced state?

Reply:
it will start on all namenodes, but will exit if it's not the active one. this is in case of HA. (if there is no active host it won't run)
if not HA it will work as it used to.

it's in cron, so it runs periodically.

I haven't touched the core logic, only the address calculation in case of HA. it was written mostly by Deveraj and you. But if we need more modification I can make it.